File: tree/101-symmetric-tree.py
# Solution 1: iterative using stack
# Time and Space complexity same as solution 1 in C++ file.
class Solution:
    def isSymmetric(self, root):
        if root is None: return True
        stack = [(root, root)]
        while stack:
            node1, node2 = stack.pop()
            if node1 is None and node2 is None: continue
            if node1 is None or node2 is None: return False
            if node1.val != node2.val: return False
            stack.append((node1.left, node2.right))
            stack.append((node1.right, node2.left))
        return True

File: practices/leetcode/tree/101-symmetric-tree.cpp
// Solution 1: recursion
// N = number of nodes
// If the tree is symmetric: (worst case)
// O(N/2) = O(N) in time
// O(2H) = O(H) in space (H = height of tree)
class Solution {
public:
    bool isSymmetric2(TreeNode* node1, TreeNode* node2) {
        if (node1 == NULL && node2 == NULL) return true;
        if (node1 == NULL || node2 == NULL) return false;
        return node1->val == node2->val &&
               isSymmetric2(node1->left, node2->right) &&
               isSymmetric2(node1->right, node2->left);
    }

    bool isSymmetric(TreeNode* root) {
        return isSymmetric2(root, root);
    }
};

// Solution 2: iterative using stack
// Please refer to solution 1 of Python file.
